### ISSUE TITLE
Multiple/parallel command execution should be prevented by ReactiveCommand

### ISSUE THREAD

**Issue by cpraehaus** [reactiveui/ReactiveUI#3022]
**Describe the bug**
``ReactiveCommand`` sets CanExecute to false as soon as the command is being executed to prevent multiple, parallel invocations. However, there are no further measures to prevent multiple/parallel executions. This behavior is inconsistent, but obviously by-design since there are unit tests for this, e.g. ``ExecuteFacilitatesAnyNumberOfInFlightExecutions``.

In particular this causes problems when you have scenarios where the user performs multiple clicks/taps on a button quickly - or while the main thread is busy/blocked: you then get e.g. 3 click events for that button. the first click causes the command to be executed. The other 2 clicks also cause command execution since CanExecute toggle to false has no effect in this case (the UI has already invoked command 3 times before actually disabling the UI button). So effectively the command is executed 3 times in parallel causing all sort of unwanted behavior.

**Steps To Reproduce**
See unit test ``ExecuteFacilitatesAnyNumberOfInFlightExecutions``. This unit test currently succeeds and exactly represents the issue discussed here: ``sub1`` and ``sub2`` both start the command in parallel. One of them "wins" and starts the command first. This causes CanExecute to toggle to false. So actually one can argue that only one of the two subscriptions effectively should cause a command execution and not both.

**Expected behaviour**

While a command is already executing further invocations of that command (via ``Execute()`` and friends) should be ignored.

I know that this might not be the expected behavior for all ReactiveUI users so one could think of making this an option to ``ReactiveCommand.Create``, e.g. *allowsMultipleExecutions*. Other frameworks have such an option (e.g. AsyncCommand in Xamarin Community Toolkit). If an option is introduced then this should also be consistent with ``CanExecute`` of course.

**Screenshots**
None.

**Environment**
- OS:  all
- Version: all
- Device: all
- ReactiveUI Version: 16.3.10

**Additional context**
None.

**Comment by tomasfil** [comment]
Hello, as far as I am aware `.Execute` doesnt respect the canExecute. Only binding command to UIElement and `.InvokeCommand` does.

**Comment by anaisbetts** [comment]
The logic is that if the programmer is explicitly calling `Execute`, they Know What They Are Doing and ReactiveCommand should not attempt to prevent it.

**Comment by cpraehaus** [comment]
@anaisbetts Execute is often not invoked by the programmer but by the framework, e.g. when you are using XAML binding (to button "Command" attribute). In this case multiple taps on the button can cause multiple invocations of the command.

If I'm required to add my own logic to prevent this whats the use of ReactiveCommand anyway?

**Comment by glennawatson** [comment]
UI frameworks call CanExecute before invoking preventing double execution

**Comment by cpraehaus** [comment]
@glennawatson Right, but the invocations happen since CanExecute does not turn false immediately:  *The other 2 clicks also cause command execution since CanExecute toggle to false has no effect in this case (the UI has already invoked command 3 times before actually disabling the UI button). So effectively the command is executed 3 times in parallel causing all sort of unwanted behavior.*

**Comment by glennawatson** [comment]
How are you producing your CanExecute observable. Sounds like a observable scheduler issue.

**Comment by cpraehaus** [comment]
@glennawatson I put together the following unit test to demonstrate what I mean:
```csharp
[Test]
public void CommandCannotBeInvokedMultipleTimes()
{
	var reactCmd = ReactiveCommand.CreateFromTask(async () => await Task.Delay(1));

	// Cast to ICommand since this the interface used by UI framework like Xamarin.Forms
	var cmd = (ICommand)reactCmd;
	Assert.That(cmd.CanExecute(null), Is.True);
	cmd.Execute(null);
	// The next Assert FAILs since CanExecute still returns true although previous
	// invocation of command is still running
	Assert.That(cmd.CanExecute(null), Is.False);
}
```

**Comment by Xo3-ToBapbl** [comment]
@cpraehaus thanks for detailed explanation. Maybe you have any advice how to overcome this issue? I'm struggling with this problem and don't fully understand how to re-write command execution invocation to prevent multiple execution by the command itself?

**Comment by anaisbetts** [comment]
@Xo3-ToBapbl Use InvokeCommand instead of Execute or check CanExecute before calling Execute

**Comment by glennawatson** [comment]
Closing this issue as part of a cleanup of long-standing issues. ReactiveCommand intentionally allows multiple in-flight executions when Execute is called directly — this is by-design and covered by unit tests. To prevent concurrent execution, use InvokeCommand (which checks CanExecute) or guard with a CanExecute check before calling Execute. Thank you for reporting it.
